File: haurylovich/compiler/syntax_analyzer.py
from antlr4 import FileStream, CommonTokenStream, InputStream
from SetLangLexer import SetLangLexer
from SetLangParser import SetLangParser
from antlr4.error.ErrorListener import ErrorListener
import sys
import os
class VerboseErrorListener(ErrorListener):
    def __init__(self):
        super(VerboseErrorListener, self).__init__()
        self.errors = []
    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        raise SyntaxError(f"Syntax error at {line}:{column}: {msg}")

def parse_from_file(file):
    input_stream = FileStream(file, encoding='utf-8')
    lexer = SetLangLexer(input_stream)
    stream = CommonTokenStream(lexer)
    parser = SetLangParser(stream)
    parser.removeErrorListeners()
    parser.addErrorListener(VerboseErrorListener())
    parser.program()
    tree = parser.program()
    return tree

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Укажите путь к файлу")
        sys.exit(1)
    file_path = sys.argv[1]
    if not os.path.isfile(file_path):
        print(f"Файл не найден: {file_path}")
        sys.exit(1)
    try:
        tree = parse_from_file(file_path)
        print("OK")
        print(tree.toStringTree(recog=None))
    except SyntaxError as e:
        print(f"ERROR: {e}")

File: haurylovich/compiler/SyntaxAnalyzer.java
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.*;
import java.io.IOException;

public class SyntaxAnalyzer {

    public static void main(String[] args) {
        String fileName = "C:\\Users\\ASUS\\IdeaProjects\\YAPIS_\\incorrect_examples\\example5.txt";

        try {
            CharStream input = CharStreams.fromFileName(fileName);
            SetLangLexer lexer = new SetLangLexer(input);
            CommonTokenStream tokens = new CommonTokenStream(lexer);
            SetLangParser parser = new SetLangParser(tokens);
            parser.removeErrorListeners();
            lexer.removeErrorListeners();
            SyntaxErrorListener errorListener = new SyntaxErrorListener();
            parser.addErrorListener(errorListener);
            lexer.addErrorListener(errorListener);
            parser.program();
            if (errorListener.hasErrors()) {
                System.out.println("Синтаксические ошибки обнаружены:");
                for (String err : errorListener.getErrors()) {
                    System.out.println(err);
                }
            } else {
                System.out.println("Синтаксис корректен!");
            }

        } catch (IOException e) {
            System.err.println("Не удалось прочитать файл: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("Ошибка при разборе: " + e.getMessage());
        }
    }

    static class SyntaxErrorListener extends BaseErrorListener {
        private final java.util.List<String> errors = new java.util.ArrayList<>();

        @Override
        public void syntaxError(Recognizer<?, ?> recognizer,
                                Object offendingSymbol,
                                int line, int charPositionInLine,
                                String msg,
                                RecognitionException e) {
            errors.add("Line " + line + ":" + charPositionInLine + " - " + msg);
        }

        public boolean hasErrors() {
            return !errors.isEmpty();
        }

        public java.util.List<String> getErrors() {
            return errors;
        }
    }
}
